test/spmd/test_xla_sharding.py

```diff
@@ -481,6 +481,41 @@ def test_xla_sharded_hlo_dump(self):
     # scalar 5 should be replicated
     self.assertIn('%p0.2 = f32[] parameter(0), sharding={replicated}', hlo)
 
+  @unittest.skipIf(xr.device_type() == 'TPU', "Crash on TPU v2")
+  @patch('torch_xla.runtime.global_device_attributes')
+  @patch('torch_xla.core.xla_model.xla_device_hw')
+  def test_2d_tensor_3d_mesh(self, xla_device_mock, device_attributes_mock):
+    xla_device_mock.return_value = "TPU"
+    device_attributes_mock.return_value = [{
+        'coords': [0, 0, 0],
+        'core_on_chip': 0
+    }, {
+        'coords': [1, 0, 0],
+        'core_on_chip': 0
+    }, {
+        'coords': [0, 1, 0],
+        'core_on_chip': 0
+    }, {
+        'coords': [1, 1, 0],
+        'core_on_chip': 0
+    }]
+    ct1 = torch.randn(16, 16, device='cpu')
+    ct2 = torch.randn(16, 16, device='cpu')
+    expected = ct1 + ct2
+
+    t1 = ct1.to(xm.xla_device())
+    t2 = ct2.to(xm.xla_device())
+    mesh = self._get_hybrid_mesh((1, self.n_devices, 1),
+                                 axis_names=('data', 'fsdp', 'tensor'))
+    t1 = xs.mark_sharding(t1, mesh, partition_spec=(0, 1))
+    if self.n_devices > 1:
+      hlo = torch_xla._XLAC._get_xla_tensors_hlo([t1.global_tensor])
+      sharding_annotation = '{devices=[1,%d,1]%s}' % (self.n_devices, ','.join(
```

Comment:
Can you share with me the actual hlo being printed? Easier for me to understand the code.

Reply:
```HloModule IrToHlo.4, is_scheduled=true, entry_computation_layout={(f32[1,4,16]{2,1,0:T(4,128)})->(f32[4,16]{1,0:T(4,128)})}

ENTRY %IrToHlo.4_spmd (param: f32[1,4,16]) -> (f32[4,16]) {
  %param = f32[1,4,16]{2,1,0:T(4,128)} parameter(0), sharding={devices=[1,4,1]0,2,1,3}
  %copy-start = (f32[1,4,16]{2,1,0:T(4,128)S(3)}, f32[1,4,16]{2,1,0:T(4,128)}, u32[]{:S(2)}) copy-start(f32[1,4,16]{2,1,0:T(4,128)} %param)
  %copy-done = f32[1,4,16]{2,1,0:T(4,128)S(3)} copy-done((f32[1,4,16]{2,1,0:T(4,128)S(3)}, f32[1,4,16]{2,1,0:T(4,128)}, u32[]{:S(2)}) %copy-start)
  %bitcast.1 = f32[4,16]{1,0:T(4,128)S(3)} bitcast(f32[1,4,16]{2,1,0:T(4,128)S(3)} %copy-done)
  %copy.1 = f32[4,16]{1,0:T(4,128)} copy(f32[4,16]{1,0:T(4,128)S(3)} %bitcast.1), backend_config="{\"flag_configs\":[],\"window_config\":{\"kernel_window_bounds\":[],\"output_window_bounds\":[\"1\",\"1\"],\"input_window_bounds\":[],\"estimated_cycles\":\"596\",\"iteration_bounds\":[\"1\",\"1\"]},\"scoped_memory_configs\":[]}"
  ROOT %tuple.2 = (f32[4,16]{1,0:T(4,128)}) tuple(f32[4,16]{1,0:T(4,128)} %copy.1)
}```

Reply:
Is it possible we keep as many of the original hlo text as possible in `sharding_annotation` such that the code is more readable?

And then paste this part as a comment to demonstrate what the end results of the string operations should look like?
```
 %param = f32[1,4,16]{2,1,0:T(4,128)} parameter(0), sharding={devices=[1,4,1]0,2,1,3}
```

It's very hard at least for me to dryrun string operations in head. So the more you show with examples the easier for reviewers to review your code and so does the further maintenance of the code.